flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartitionTest.java

```diff
@@ -347,6 +357,129 @@ public void testProducerFailedException() {
         assertTrue(view.getFailureCause() instanceof CancelTaskException);
     }
 
+    @Test
+    public void testConsumeTimeoutableCheckpointBarrierQuickly() throws Exception {
+        PipelinedSubpartition subpartition = createSubpartition();
+        subpartition.setChannelStateWriter(ChannelStateWriter.NO_OP);
+        assertEquals(0, subpartition.getChannelStateFutures().size());
+        assertEquals(0, subpartition.getNumberOfQueuedBuffers());
+
+        // test without data buffer
+        long checkpointId = 5L;
+        subpartition.add(getTimeoutableBarrierBuffer(checkpointId));
+        assertEquals(1, subpartition.getNumberOfQueuedBuffers());
+        assertEquals(1, subpartition.getChannelStateFutures().size());
+        CompletableFuture<List<Buffer>> channelStateFuture =
+                subpartition.getChannelStateFutures().get(checkpointId);
+        assertFalse(channelStateFuture.isDone());
+
+        ResultSubpartition.BufferAndBacklog barrierBuffer = subpartition.pollBuffer();
+        assertNotNull(barrierBuffer);
+        assertEquals(
+                Buffer.DataType.TIMEOUTABLE_ALIGNED_CHECKPOINT_BARRIER,
+                barrierBuffer.buffer().getDataType());
+
+        assertEquals(0, subpartition.getChannelStateFutures().size());
+        assertEquals(0, subpartition.getNumberOfQueuedBuffers());
+        assertNull(channelStateFuture.get());
+
+        // test with data buffer
+        subpartition.resumeConsumption();
+        checkpointId = 6;
+        subpartition.add(createFilledFinishedBufferConsumer(4096));
+        subpartition.add(getTimeoutableBarrierBuffer(checkpointId));
+        assertEquals(1, subpartition.getChannelStateFutures().size());
+        assertEquals(2, subpartition.getNumberOfQueuedBuffers());
+        channelStateFuture = subpartition.getChannelStateFutures().get(checkpointId);
+
+        assertNotNull(subpartition.pollBuffer());
+        assertFalse(channelStateFuture.isDone());
+        barrierBuffer = subpartition.pollBuffer();
+        assertNotNull(barrierBuffer);
+        assertEquals(
+                Buffer.DataType.TIMEOUTABLE_ALIGNED_CHECKPOINT_BARRIER,
+                barrierBuffer.buffer().getDataType());
+
+        assertEquals(0, subpartition.getChannelStateFutures().size());
+        assertEquals(0, subpartition.getNumberOfQueuedBuffers());
+        assertNull(channelStateFuture.get());
+    }
```

Comment:
Maybe write a helper method `testWithNDataBuffers(int n, ...)`; and then call it twice:
```
testWithNDataBuffers(0, ...);
testWithNDataBuffers(1, ...);
```
to deduplicate the code?